src/push/apns.rs

```diff
@@ -68,7 +70,19 @@ pub fn send_push(
             .send(payload)
             .map(|response| debug!("Success details: {:?}", response))
             .map_err(|error| {
-                SendPushError::ProcessingError(format!("Push was unsuccessful: {}", error))
+                // Treat "bad device token" errors as client errors
+                if let A2Error::ResponseError(ref resp) = error {
+                    if let Some(ref body) = resp.error {
+                        if body.reason == ErrorReason::BadDeviceToken {
+                            return SendPushError::ProcessingClientError(
+                                format!("Push was unsuccessful: Bad device token")
+                            );
+                        }
+                    }
+                }
+
+                // Treat all other errors as server errors
+                SendPushError::ProcessingServerError(format!("Push was unsuccessful: {}", error))
```

Comment:
What does the server return if it is too busy? In that case, either the relay server could delay the response or the client detect that the server is busy (by a specific status code range) and slow down its requests.

Reply:
Probably a [`ConnectionError`](https://pimeys.github.io/a2/v0.3.3/a2/error/enum.Error.html). That could have different reasons though.

But in the end, as long as we retry pushes (after some delay) in the client on server (but not client) errors, we should be fine.